psp/ml/models/recent_history.py

```diff
@@ -0,0 +1,190 @@
+import dataclasses
+from collections import defaultdict
+from datetime import datetime, timedelta
+from typing import Iterator, Tuple
+
+import numpy as np
+import xarray as xr
+
+from psp.ml.dataset import PvDataSource
+from psp.ml.models.base import PvSiteModel, PvSiteModelConfig
+from psp.ml.models.regressors.base import Regressor
+from psp.ml.typings import Batch, Features, X, Y
+from psp.pv import get_irradiance
+from psp.utils.maths import safe_div
+
+
+def to_midnight(ts: datetime) -> datetime:
+    return ts.replace(hour=0, minute=0, second=0, microsecond=0)
+
+
+def tod_as_minutes(ts: datetime) -> float:
```

Comment:
`tod` = time of day, is worth ust calling it `time` or `time_of_day`. I find some abbreviations less helpful